mmdet/utils/memory.py

```diff
@@ -0,0 +1,172 @@
+# Copyright (c) OpenMMLab. All rights reserved.
+# Modified from https://github.com/facebookresearch/detectron2/blob/main/detectron2/utils/memory.py  # noqa
+import warnings
+from contextlib import contextmanager
+from functools import wraps
+
+import torch
+
+from mmdet.utils import get_root_logger
+
+
+@contextmanager
+def _ignore_torch_cuda_oom():
+    """A context which ignores CUDA OOM exception from pytorch."""
+    try:
+        yield
+    except RuntimeError as e:
+        # NOTE: the string may change?
+        if 'CUDA out of memory. ' in str(e):
+            pass
+        else:
+            raise
+
+
+class AvoidOOM(object):
+    """Try to convert inputs to FP16 and CPU if got a PyTorch's CUDA Out of
+    Memory error.
+
+    Args:
+        logger (logging.Logger | str, optional): Logger used for printing
+            related information during evaluation. Default: None.
+        return_gpu (bool): Whether convert outputs back to GPU. Default: True.
+        out_fp32 (bool): Whether to convert the output back to out_fp32.
+            Default: True.
+        output_type (torch.dtype): Destination type. Default: None
+
+    Examples:
+        >>> from mmdet.utils.memory import AvoidOOM
+        >>> AvoidOOM = AvoidOOM()
+        >>> output = AvoidOOM.retry_if_cuda_oom(
+        >>>     some_torch_function)(input1, input2)
+
+    Note:
+        1. The output may be on CPU even if inputs are on GPU. Processing
+            on CPU will slows down the code significantly.
+        2. When converting inputs to CPU, it will only look at each argument
+            and check if it has `.device` and `.to` for conversion. Nested
+            structures of tensors are not supported.
+        3. Since the function might be called more than once, it has to be
+            stateless.
+    """
+
+    def __init__(self,
+                 logger=None,
+                 out_fp32=True,
+                 return_gpu=True,
+                 output_type=None,
+                 convert_cpu=True):
+        self.logger = logger
+        self.return_gpu = return_gpu
+        if output_type is not None:
```

Comment:
I think this is determined by the input type rather than the config. We should ensure that tensors have the same type when they are passed in and processed out.